### ISSUE TITLE
[False Positive][ interprocedural ] is_null_checked  test is not that robust

### ISSUE THREAD

**Issue by GeekMin** [spotbugs/spotbugs#95]
`
String uin = null
...
if (android.text.TextUtils.isEmpty(uin)) { 
   uin.length();
    //bla-bla
} else {
  // bla-bla
}
`

here, inside TextUtils.isEmpty, uin is null-checked. 
because of findbugs' interprocedural is not that robust, FP generated...

Here, how can I solve this problem..?

(Like Coverity, you can provide one model(Klocwork calls  it knowledge)(to tell the tool something like this:
`package android.text;
import static com.coverity.primitives.Coverity.*;
import static com.coverity.modeling.Helpers.*;

public class TextUtils {
    public static void getChars(CharSequence s, int start, int end, char[] dest, int destoff) {
        if(start < 0) {
            killpath();
        }
        deref(dest);
		
        com.coverity.primitives.Dataflow.dataflow_param_merge(dest, s); // WARNING: unsupported model primitive
    }
	static boolean isEmpty(CharSequence str){
		if (str == null or len(str) ==0)
			return true;
		else:
			return false;
	}
}
`
then, problems solved..

so, If I use findbugs(spotbugs), I should dive into source code to write a custom checker or modify source code in-place to tell the tool: hi, here, is null-checked..?

**Comment by GeekMin** [comment]
what I think is like this:
I can give the tool some knowledge:
TextUtils.isEmpty(arg)  <-> if (arg == null)
!TextUtils.isEmpty(arg) <-> if (arg != null)

**Comment by jsotuyod** [comment]
Hi! I'm a little bit confused as to your report. I'm not sure what bug description you refer to by *"is_null_checked"*. Can you please state exactly the bug being reported by Findbugs / Spotbugs?

As you say, Android's `TextUtils.isEmpty` return true for both null and empty string. Therefore, a code such as:

```
String uin = null;
if (android.text.TextUtils.isEmpty(uin)) {
    uin.length(); // This produces a Null Pointer Exception!
}
```

As such, if it is reporting a [null pointer dereference](https://spotbugs.github.io/bugDescriptions.html#NP_ALWAYS_NULL), then the behavior is actually correct.

Could you please clarify? Thanks in advanced

**Comment by GeekMin** [comment]
`class TestTextUtils {
    public void foo(String uin){
        if (android.text.TextUtils.isEmpty(uin)){
            System.out.println("do nothing");
            uin.length();  // expected warning here
        } else {
            uin.length();
        }
    }
        
    static boolean isEmpty(String uin){
        if (uin == null || uin.length()==0)
            return true;
        else
            return false;
            
    }
    public void bar(String uin){
        if (isEmpty(uin)){
            System.out.println("do nothing");
            uin.length();   // expected warning here
        } else {
            uin.length();
        }
        
    }
}`

sorry, previous example is not that correct..  you can refer to below code..
Checking this code, Findbugs doesn't find anything..
